src/GenSymIO.chpl

```diff
@@ -1928,146 +1949,101 @@ module GenSymIO {
-        var indices: list(int);
-        var i: int = 0;
-        indices.append(0);   
-        var segmentsBound = charList.size - sliceIndex - 1;
-
-        for value in charList(sliceIndex..charList.size-1)  {
-            valuesList.append(value:uint(8));
-            
-            /*
-             * If the value is a null char and is not the last char
-             * in the list, then it is a segment use to delimit
-             * strings within the corresponding values array.
-             */
-            if value == NULL_STRINGS_VALUE && i < segmentsBound {
-                indices.append(i+1);
-            }
-            i+=1;
-        }
-
-        return (valuesList,indices);
+    private proc adjustForLeftShuffle(shuffleLeftIndex: int, charList, idx: int) throws {
+        return charList[shuffleLeftIndex..charList.size-1];
     }
 
     /* 
-     * Adjusts for the shuffling of a trailing char sequence to the next 
-     * locale by (1) slicing trailing chars that correspond to 1..n
-     * chars composing a string that completes in the next locale,
-     * returning (1) a new list that composes all strings that end in the 
-     * current locale and (2) returns a new segments list corresponding
-     * to the new values list for the current locale
+     * Adjusts for the right shuffle of the trailing char sequence from the current locale
+     * to the next locale by returning a slice containing chars up to and including 
+     * the rightShuffleIndex. 
      */
-    private proc adjustForTrailingSlice(sliceIndex : int,
-                                   charList : list(uint(8))) throws {
-        var valuesList: list(uint(8), parSafe=false);
-        var indices: list(int);
-        var i: int = 0;
-        indices.append(0);
-
-        for value in charList(0..sliceIndex-1)  {
-            valuesList.append(value:uint(8));
-            if value == NULL_STRINGS_VALUE && i < sliceIndex-1 {
-                indices.append(i+1);
+    private proc adjustForRightShuffle(shuffleRightIndex: int, 
+                                               charsList: list(uint(8)), idx : int) throws {        
+        return charsList[0..shuffleRightIndex];
+    }
+
+    private proc generateFinalSegmentsList(charList : list(uint(8)), idx: int) throws {
+        var segments: list(int);
+        segments.append(0);
+
+        for (value, i) in zip(charList, 0..charList.size-1) do {
+            /*
+             * If the char is the null uint(8) char, check to see if it is the 
+             * last char. If not, added to the indices. If it is the last char,  
+             * don't add, because it is the correct ending char for a Strings 
+             * values array to be written to a locale.
+             */ 
+            if value == NULL_STRINGS_VALUE && i < charList.size-1 {
+                segments.append(i+1);
             }
-            i+=1;
         }
-
-        return (valuesList,indices);
+        
+        return segments;
```

Comment:
Man this stinks that we have to scan over the bytes themselves looking for Nulls to generate the segments when in theory we already know where all the nulls are based on the offsets/segments pdarray... but I get it, there is a lot of moving around going on and those segments are being localized / re-indexed etc.  So this entire PR is a great  step in streamlining & making things faster.

Reply:
Yes, as soon as chars get shuffled around, the segments have to change.